File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/OrangesRotting.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayDeque;
import java.util.Queue;

public class OrangesRotting {

    private static final int[][] DIRECTIONS = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
    private static final int EMPTY = 0;
    private static final int FRESH = 1;
    private static final int ROTTEN = 2;

    public int orangesRotting(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }

        int rows = grid.length;
        int cols = grid[0].length;
        int freshCount = 0;
        Queue<int[]> rottenQueue = new ArrayDeque<>();
        int minutes = 0;

        // 统计新鲜橘子数量并将腐烂橘子加入队列
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == FRESH) {
                    freshCount++;
                } else if (grid[row][col] == ROTTEN) {
                    rottenQueue.offer(new int[]{row, col});
                }
            }
        }

        // 如果没有新鲜橘子，直接返回0
        if (freshCount == 0) {
            return 0;
        }

        // BFS模拟腐烂过程
        while (!rottenQueue.isEmpty() && freshCount > 0) {
            int queueSize = rottenQueue.size();

            // 处理当前分钟所有腐烂的橘子
            for (int i = 0; i < queueSize; i++) {
                int[] current = rottenQueue.poll();
                int currentRow = current[0];
                int currentCol = current[1];

                // 检查四个方向
                for (int[] direction : DIRECTIONS) {
                    int newRow = currentRow + direction[0];
                    int newCol = currentCol + direction[1];

                    if (isValidPosition(rows, cols, newRow, newCol) &&
                            grid[newRow][newCol] == FRESH) {
                        grid[newRow][newCol] = ROTTEN;
                        freshCount--;
                        rottenQueue.offer(new int[]{newRow, newCol});
                    }
                }
            }
            minutes++;
        }

        return freshCount == 0 ? minutes : -1;
    }

    private boolean isValidPosition(int rows, int cols, int row, int col) {
        return row >= 0 && row < rows && col >= 0 && col < cols;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/orangesRotting.kt
package com.only4.algorithm.version4k.leetcode

private const val EMPTY = 0
private const val FRESH = 1
private const val ROTTEN = 2

fun orangesRotting(grid: Array<IntArray>): Int {
    if (grid.isEmpty() || grid[0].isEmpty()) return 0

    val rows = grid.size
    val cols = grid[0].size
    val directions = arrayOf(Pair(-1, 0), Pair(1, 0), Pair(0, -1), Pair(0, 1))
    val rottenQueue = ArrayDeque<Pair<Int, Int>>()
    var freshCount = 0
    var minutes = 0

    fun isValidPosition(row: Int, col: Int) = row in 0 until rows && col in 0 until cols

    // 统计新鲜橘子数量并将腐烂橘子加入队列
    for (row in 0 until rows) {
        for (col in 0 until cols) {
            when (grid[row][col]) {
                FRESH -> freshCount++
                ROTTEN -> rottenQueue.add(Pair(row, col))
            }
        }
    }

    // 如果没有新鲜橘子，直接返回0
    if (freshCount == 0) return 0

    // BFS模拟腐烂过程
    while (rottenQueue.isNotEmpty() && freshCount > 0) {
        // 处理当前分钟所有腐烂的橘子
        repeat(rottenQueue.size) {
            val (currentRow, currentCol) = rottenQueue.removeFirst()

            // 检查四个方向的相邻位置
            for ((deltaRow, deltaCol) in directions) {
                val newRow = currentRow + deltaRow
                val newCol = currentCol + deltaCol

                // 检查新坐标是否在网格内且是新鲜橘子
                if (isValidPosition(newRow, newCol) && grid[newRow][newCol] == FRESH) {
                    // 将新鲜橘子标记为腐烂
                    grid[newRow][newCol] = ROTTEN
                    freshCount--
                    rottenQueue.add(Pair(newRow, newCol))
                }
            }
        }
        minutes++
    }

    return if (freshCount == 0) minutes else -1
}

